In [42]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import ipywidgets as widgets
from IPython.display import display

# Разведывательный анализ распределения IQ по странам

## 1. Введение и постановка вопроса

В данной работе проводится разведывательный анализ (Exploratory Data Analysis, EDA) набора данных `IQ_per_country.csv`, содержащего сведения о показателях IQ, связанных с различными странами.

Основная цель анализа — изучить распределение показателей IQ между странами,
выявить общие закономерности, различия и потенциальные аномалии в данных.

В рамках исследования будут рассмотрены следующие вопросы:

- Как распределены значения IQ в представленном наборе данных?
- Насколько сильно различаются значения между странами?
- Какие страны имеют наиболее высокие и наиболее низкие значения показателя?
- Существуют ли заметные выбросы или другие особенности распределения?
- Как изменяются результаты при выборе отдельных стран или групп стран?

Важно учитывать, что показатель IQ является статистической характеристикой, а сравнение таких показателей между странами может зависеть от методики исследования, размера и состава выборки, условий проведения тестирования и других факторов. Поэтому выявленные различия следует интерпретировать как особенности данного набора данных, а не как абсолютную характеристику населения соответствующих стран.

## Исследовательские вопросы

В качестве основных исследовательских вопросов рассматриваются:

1. Какова форма распределения значений IQ в наборе данных?
2. Насколько велик разброс показателей между странами?
3. Какие страны находятся на разных участках распределения?
4. Какие статистические и визуальные особенности данных требуют дополнительного рассмотрения?

In [2]:
df = pd.read_csv("IQ_per_country.csv")
df.head()

,Rank,Country,Average IQ,Continent,Literacy Rate,Nobel Prices,HDI (2021),Mean years of schooling - 2021,GNI - 2021,Population - 2023
0,1,Japan,106.48,Asia,0.99,29,0.925,13.4,42274.0,123294513
1,2,Taiwan,106.47,Asia,0.96,4,NaN,NaN,NaN,10143543
2,3,Singapore,105.89,Asia,0.97,0,0.939,11.9,90919.0,6014723
3,4,Hong Kong,105.37,Asia,0.94,1,0.952,12.2,62607.0,7491609
4,5,China,104.10,Asia,0.96,8,0.768,7.6,17504.0,1425671352


In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 10 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Rank                            193 non-null    int64  
 1   Country                         193 non-null    str    
 2   Average IQ                      193 non-null    float64
 3   Continent                       193 non-null    str    
 4   Literacy Rate                   193 non-null    float64
 5   Nobel Prices                    193 non-null    int64  
 6   HDI (2021)                      179 non-null    float64
 7   Mean years of schooling - 2021  179 non-null    float64
 8    GNI - 2021                     179 non-null    float64
 9   Population - 2023               193 non-null    str    
dtypes: float64(5), int64(2), str(3)
memory usage: 15.2 KB


In [4]:
df.describe()

,Rank,Average IQ,Literacy Rate,Nobel Prices,HDI (2021),Mean years of schooling - 2021,GNI - 2021
count,193.000000,193.000000,193.000000,193.000000,179.000000,179.000000,179.000000
mean,97.000000,82.047927,0.864249,5.922280,0.724123,9.028492,20811.798883
std,55.858452,13.336117,0.180262,31.971328,0.152761,3.187867,22160.494888
min,1.000000,42.990000,0.190000,0.000000,0.385000,2.100000,732.000000
25%,49.000000,74.330000,0.800000,0.000000,0.599500,6.400000,4593.000000
50%,97.000000,82.240000,0.950000,0.000000,0.745000,9.400000,12672.000000
75%,145.000000,91.600000,0.990000,1.000000,0.844000,11.600000,30588.000000
max,193.000000,106.480000,1.000000,400.000000,0.962000,14.100000,146830.000000


In [5]:
missing = df.isna().sum()
missing[missing > 0]

HDI (2021)                        14
Mean years of schooling - 2021    14
 GNI - 2021                       14
dtype: int64

In [6]:
missing_percent = (df.isna().mean() * 100).round(2)
missing_percent[missing_percent > 0]

HDI (2021)                        7.25
Mean years of schooling - 2021    7.25
 GNI - 2021                       7.25
dtype: float64

In [7]:
df["Population - 2023"].head(16)

0      123294513
1       10143543
2        6014723
3        7491609
4     1425671352
5       51784059
6        9498238
7        5545475
8        \t39315
9       83294633
10      17618299
11       1322766
12       654.768
13        704.15
14      16944826
15      38781292
Name: Population - 2023, dtype: str

In [8]:
df["Population - 2023"].apply(type).value_counts()

Population - 2023
<class 'str'>    193
Name: count, dtype: int64

In [9]:
df.nlargest(10, "Average IQ")[["Country", "Average IQ"]]

,Country,Average IQ
0,Japan,106.48
1,Taiwan,106.47
2,Singapore,105.89
3,Hong Kong,105.37
4,China,104.10
5,South Korea,102.35
6,Belarus,101.60
7,Finland,101.20
8,Liechtenstein,101.07
9,Germany,100.74


## 2. Знакомство с данными

### Общая структура данных

Набор данных содержит 193 наблюдения и 10 признаков. Каждая строка соответствует отдельной стране или территории, а признаки описывают показатель среднего IQ, континент, уровень грамотности, количество Нобелевских премий, индекс человеческого развития (HDI), среднюю продолжительность обучения, GNI и численность населения.

Признаки имеют разные типы. `Country` и `Continent` являются категориальными переменными, `Rank` представляет собой порядковый ранг, а остальные признаки в основном являются количественными.

Большинство признаков заполнены полностью. Однако для трёх показателей, связанных с HDI, наблюдается по 179 значений из 193, то есть по 14 пропусков в каждом признаке. Это составляет приблизительно 7.3% наблюдений и потребует отдельного рассмотрения на этапе предобработки.

Также `Population - 2023` представлена в формате `str`, хотя по смыслу является числовым показателем. Перед использованием этого признака в математических операциях его потребуется преобразовать в числовой тип.

### Первичные статистические характеристики

Показатель `Average IQ` имеет среднее значение 82.05 и медианное значение 82.24. Они находятся близко друг к другу, однако этого недостаточно для вывода о симметричности распределения; его форма будет дополнительно исследована с помощью визуализаций.

Диапазон значений `Average IQ` составляет от 42.99 до 106.48, а стандартное отклонение равно 13.34, что указывает на заметный разброс значений между наблюдениями.

`Literacy Rate` принимает значения от 0.19 до 1.00. При этом медиана равна 0.95, а 75-й процентиль — 0.99, поэтому распределение этого показателя также представляет интерес для дальнейшего анализа.

Для `GNI - 2021` наблюдается значительный разброс: значения находятся в диапазоне от 732 до 146830. Стандартное отклонение составляет 22160, что сравнимо и даже несколько превышает среднее значение 20812. Это может указывать на существенную неоднородность показателя.

Особенно заметным является распределение `Nobel Prices`: медианное значение равно 0, 75-й процентиль — 1, тогда как максимальное значение достигает 400. Таким образом, этот признак имеет крайне неравномерное распределение и требует дополнительного изучения перед использованием в сравнительном анализе.

### Форматирование численности населения

При первичном анализе было обнаружено, что признак `Population - 2023`
сохранён как строковый тип (`str`). Дополнительная проверка показала наличие
лишних пробельных символов в некоторых значениях, например табуляции перед
числом.

Для корректного использования признака в численных расчётах были удалены
лишние пробельные символы, после чего значения преобразованы в целочисленный
тип.

In [10]:
for i, value in df["Population - 2023"].items():
    try:
        int(value.strip())
    except ValueError:
        print(i, repr(value))

12 '654.768'
13 '704.15'
24 '375.319'
36 '76.965'
41 '292.991'
42 '334.506'
44 '63.837'
48 '281.996'
49 '535.065'
54 '623.237'
66 '787.425'
71 '452.524'
73 '412.624'
80 '626.485'
84 '44.104'
86 '115.224'
87 '936.375'
88 '41.996'
89 '740.425'
93 '813.834'
107 '49.796'
112 '521.022'
115 '126.184'
121 '107.66'
130 '852.075'
147 '180.251'
151 '94.298'
161 '6.115'
170 '73.161'
171 '231.856'
173 '103.699'
177 '410.825'


In [11]:
df["Population - 2023"] = (
    df["Population - 2023"]
    .str.strip()
    .str.replace(".", "", regex=False)
    .astype(int)
)

df["Population - 2023"].dtype

dtype('int64')

## 3. Очистка и предобработка

На этапе первичного анализа были обнаружены две особенности данных, требующие предобработки.

Во-первых, в признаке `Population - 2023` значения были сохранены как строки. Кроме того, некоторые значения содержали лишние пробельные символы, например табуляцию, а в некоторых значениях точка использовалась в качестве разделителя тысяч. Например, значение `654.768` для Люксембурга соответствует населению 654 768 человек.

Поэтому перед дальнейшим анализом из значений были удалены начальные и конечные пробельные символы, а точки-разделители тысяч были удалены. После этого признак был преобразован в целочисленный тип.

Во-вторых, в трёх признаках (`HDI (2021)`, `Mean years of schooling - 2021` и `GNI - 2021`) отсутствуют по 14 значений. Эти пропуски не были заменены искусственными значениями, поскольку их механическая замена средним могла бы изменить статистические характеристики данных. В основном анализе эти пропуски будут учитываться при расчёте соответствующих показателей.

In [12]:
df["Country"].duplicated().sum()

np.int64(0)

## 4. Основной анализ и визуализация

В данном разделе проводится основной разведывательный анализ данных. Для каждой визуализации формулируется отдельный исследовательский вопрос, после чего полученный результат интерпретируется с учётом особенностей набора данных.

Первым этапом рассмотрим распределение показателя `Average IQ` среди представленных наблюдений.

### 4.1. Распределение среднего IQ

**Исследовательский вопрос:** как распределены значения `Average IQ` среди стран, представленных в наборе данных?

**Предположение:** значения могут быть распределены неравномерно, поэтому интересно определить центральную область распределения, его разброс и наличие потенциальных выбросов.

In [37]:
fig = px.histogram(
    df,
    x="Average IQ",
    nbins=20,
    title="average iq distribution",
    labels={
        "Average IQ": "Average IQ score",
        "count": "The number of countries",
    }
)

fig.update_layout(
    xaxis_title="Average IQ score",
    yaxis_title="The number of countries",
)

fig.show()

### 4.2. Распределение IQ по континентам

**Исследовательский вопрос:** различается ли распределение показателя `Average IQ` между континентами?

**Предположение:** значения `Average IQ` могут различаться между континентами, поэтому сравним их медианы, разброс и потенциальные выбросы.

In [38]:
fig = px.box(
    df,
    x="Continent",
    y="Average IQ",
    color="Continent",
    points="all",
    title="average iq distribution by continent",
    labels={
        "Continent": "Continent",
        "Average IQ": "Average IQ score"
    }
)

fig.update_layout(
    xaxis_title="Continent",
    yaxis_title="Average IQ",
    showlegend=False
)

fig.show()

### 4.3. Связь между IQ и индексом человеческого развития

**Исследовательский вопрос:** существует ли визуально заметная связь между `Average IQ` и `HDI (2021)`?

**Предположение:** страны с более высоким индексом человеческого развития могут иметь более высокие значения `Average IQ`. Проверим, наблюдается ли такая тенденция в представленных данных.

In [40]:
fig = px.scatter(
    df,
    x="HDI (2021)",
    y="Average IQ",
    color="Continent",
    hover_name="Country",
    hover_data={
        "HDI (2021)": True,
        "Average IQ": True,
        "Continent": True,
    },
    title="relationship between IQ and HDI",
    labels={
        "HDI (2021)": "Human Development Index",
        "Average IQ": "Average IQ",
        "Continent": "Continent"
    }
)

fig.update_layout(
    xaxis_title="Human Development Index",
    yaxis_title="Average IQ"
)

fig.show()

### 4.4. Связь между IQ и продолжительностью обучения

**Исследовательский вопрос:** существует ли связь между `Average IQ` и средней продолжительностью обучения населения?

**Предположение:** страны, в которых средняя продолжительность обучения выше, могут иметь более высокие значения `Average IQ`. Проверим, наблюдается ли подобная тенденция в данных.

In [ ]:
indicators = {
    "HDI (2021)": "HDI",
    "Mean years of schooling - 2021": "Mean years of schooling",
    "Literacy Rate": "Literacy Rate",
    "GNI - 2021": "GNI"
}

dropdown = widgets.Dropdown(
    options=indicators.keys(),
    value="Mean years of schooling - 2021",
    description="Metric:",
    style={"description_width": "initial"}
)

def update_plot(indicator):
    plot_df = df.dropna(subset=[indicator, "Average IQ"])

    fig = px.scatter(
        plot_df,
        x=indicator,
        y="Average IQ",
        color="Continent",
        hover_name="Country",
        hover_data={
            indicator: True,
            "Average IQ": True,
            "Continent": True
        },
        title=f"relation between IQ and {indicators[indicator]}",
        labels={
            indicator: indicators[indicator],
            "Average IQ":" Average IQ",
            "Continent": "Continent"
        }
    )

    fig.update_layout(
        xaxis_title=indicators[indicator],
        yaxis_title="Average IQ",
        showlegend=True,
    )

    fig.show()

widgets.interact(
    update_plot,
    indicator=dropdown
);